### ISSUE TITLE
tiny TUYA: NameError: name 'last_update' is not defined

### ISSUE THREAD

**Issue by DarkVillager** [Xenomes/Domoticz-TinyTUYA-Plugin#38]
I just installed tiny tuya. I followed the steps & my connection seems sucessesful because I get 0 errors about the connection.

```bat
2023-04-14 12:25:30.091 Error: tiny TUYA: Call to function 'onHeartbeat' failed, exception details:
2023-04-14 12:25:30.096 Error: tiny TUYA: Traceback (most recent call last):
2023-04-14 12:25:30.096 Error: tiny TUYA: File "/home/pi/domoticz/plugins/Domoticz-TinyTUYA-Plugin/plugin.py", line 321, in onHeartbeat
2023-04-14 12:25:30.096 Error: tiny TUYA: _plugin.onHeartbeat()
2023-04-14 12:25:30.096 Error: tiny TUYA: File "/home/pi/domoticz/plugins/Domoticz-TinyTUYA-Plugin/plugin.py", line 282, in onHeartbeat
2023-04-14 12:25:30.096 Error: tiny TUYA: if time.time() - last_update < 60 and testData == False:
2023-04-14 12:25:30.096 Error: tiny TUYA: NameError: name 'last_update' is not defined
```

[Full bin](https://hastebin.com/share/badapegice.yaml)

**Comment by Xenomes** [comment]
You are using a old version
`2023-04-14 12:25:21.938 Status: tiny TUYA: Initialized version 1.3.9, author 'Xenomes'`
Current version is 1.5.1

**Comment by DarkVillager** [comment]
> You are using a old version
> `2023-04-14 12:25:21.938 Status: tiny TUYA: Initialized version 1.3.9, author 'Xenomes'`
> Current version is 1.5.1

oh i see. Will try to update.

**Comment by RolandTwilt** [comment]
> You are using a old version `2023-04-14 12:25:21.938 Status: tiny TUYA: Initialized version 1.3.9, author 'Xenomes'` Current version is 1.5.1

Same person as @DarkVillager 

Now we get the following error:
```bat
2023-04-14 15:01:03.055 Error: tiny TUYA: handleThread: string indices must be integers
```

More detailed (debug: all):
```bat
2023-04-14 15:03:03.120 Error: tiny TUYA: handleThread: string indices must be integers
2023-04-14 15:03:03.121 Error: tiny TUYA: Call to function 'onStop' failed, exception details:
2023-04-14 15:03:03.129 Error: tiny TUYA: Traceback (most recent call last):
2023-04-14 15:03:03.129 Error: tiny TUYA: File "/home/pi/domoticz/plugins/Domoticz-TinyTUYA-Plugin/plugin.py", line 353, in onStop
2023-04-14 15:03:03.129 Error: tiny TUYA: _plugin.onStop()
2023-04-14 15:03:03.129 Error: tiny TUYA: File "/home/pi/domoticz/plugins/Domoticz-TinyTUYA-Plugin/plugin.py", line 101, in onStop
2023-04-14 15:03:03.129 Error: tiny TUYA: if Devices[dev['id']].Units[1].sValue == 'This device is not reconised, edit and run the debug_discovery with python from the tools directory and receate a issue report at https://github.com/Xenomes/Domoticz-TinyTUYA-Plugin/issues so the device can be added.':
2023-04-14 15:03:03.129 Error: tiny TUYA: TypeError: string indices must be integers
```

**Comment by Xenomes** [comment]
onStop command. can you change to the develop branch, i have made a update in there.

**Comment by RolandTwilt** [comment]
> onStop command. can you change to the develop branch, i have made a update in there.

Changed auth token to (auth token here) (for safety ig)

```bat
2023-04-14 15:17:00.552 Error: tiny TUYA: handleThread: string indices must be integers
2023-04-14 15:17:00.555 Error: tiny TUYA: Call to function 'onStop' failed, exception details:
2023-04-14 15:17:00.565 Error: tiny TUYA: Traceback (most recent call last):
2023-04-14 15:17:00.565 Error: tiny TUYA: File "/home/pi/domoticz/plugins/Domoticz-TinyTUYA-Plugin/plugin.py", line 357, in onStop
2023-04-14 15:17:00.565 Error: tiny TUYA: _plugin.onStop()
2023-04-14 15:17:00.565 Error: tiny TUYA: File "/home/pi/domoticz/plugins/Domoticz-TinyTUYA-Plugin/plugin.py", line 101, in onStop
2023-04-14 15:17:00.565 Error: tiny TUYA: if str(Devices[dev['id']].Units[1].sValue) == 'This device is not reconised, edit and run the debug_discovery with python from the tools directory and receate a issue report at https://github.com/Xenomes/Domoticz-TinyTUYA-Plugin/issues so the device can be added.':
2023-04-14 15:17:00.565 Error: tiny TUYA: TypeError: string indices must be integers
2023-04-14 15:17:00.763 Error: CheckAuthToken(<auth token here>) : auth token mismatch

**Comment by RolandTwilt** [comment]
Also got this from running debug_discovery.py
```py
pi@raspberrypi:~/domoticz/plugins/Domoticz-TinyTUYA-Plugin/tools $ python debug_discovery.py
No device data returnd for Tuya. Trying again!
```

**Comment by Xenomes** [comment]
What is the tintuya version? of the user and root?

**Comment by RolandTwilt** [comment]
> What is the tintuya version? of the user and root?

1.12.3

```bat
pi@raspberrypi:~/domoticz/plugins/Domoticz-TinyTUYA-Plugin/tools $ sudo pip show tinytuya
Name: tinytuya
Version: 1.12.3
Summary: Python module to interface with Tuya WiFi smart devices
Home-page: https://github.com/jasonacox/tinytuya
Author: Jason Cox
Author-email: jason@jasonacox.com
License: UNKNOWN
Location: /usr/local/lib/python3.7/dist-packages
Requires: colorama, pycryptodome, requests
Required-by: 
```

**Comment by Xenomes** [comment]
can you try debug_discovery.py with sudo

**Comment by RolandTwilt** [comment]
> 

Lemme check

**Comment by RolandTwilt** [comment]
Same

```bat
pi@raspberrypi:~/domoticz/plugins/Domoticz-TinyTUYA-Plugin/tools $ sudo python debug_discovery.py
No device data returnd for Tuya. Trying again!
debug_discovery: -1 line 50


dump.json is created!
```

**Comment by Xenomes** [comment]
I am not home right now, i will do some tests tonight. It looks like there is no data for tinytuya or tuya it self.

**Comment by RolandTwilt** [comment]
> I am not home right now, i will do some tests tonight. It looks like there is no data for tinytuya or tuya it self.

Tuya works via e.g. getting a device via the browser but tinytuya doesn't I think.

**Comment by Xenomes** [comment]
Can you run python3 -m tinytuya wizard

**Comment by RolandTwilt** [comment]
> Can you run python3 -m tinytuya wizard

I get this?

```bat
Error from Tuya server: Error from Tuya Cloud: Code 28841002: 'No permissions. Your subscription to cloud development plan has expired.'
Check DeviceID and Region
```

**Comment by Xenomes** [comment]
you can check 'cloud development plan has expired' that?

**Comment by RolandTwilt** [comment]
> you can check 'cloud development plan has expired' that?

how

**Comment by RolandTwilt** [comment]
![image](https://user-images.githubusercontent.com/130754579/232060548-c02df74d-d1cb-4397-afe8-958b8184a8ab.png)

**Comment by Xenomes** [comment]
https://eu.iot.tuya.com/cloud/products/detail
Is your subscription to cloud development plan expired, you can extend it <a href="https://iot.tuya.com/cloud/products/apply-extension"> HERE</a><br/>

**Comment by RolandTwilt** [comment]
> https://eu.iot.tuya.com/cloud/products/detail Is your subscription to cloud development plan expired, you can extend it [ HERE](https://iot.tuya.com/cloud/products/apply-extension)

i will try that.

**Comment by RolandTwilt** [comment]
> https://eu.iot.tuya.com/cloud/products/detail Is your subscription to cloud development plan expired, you can extend it [ HERE](https://iot.tuya.com/cloud/products/apply-extension)

I got the extension but still the same error even tho python3 -m tinytuya wizard
works

**Comment by Xenomes** [comment]
Can you update to version 1.5.3

**Comment by RolandTwilt** [comment]
> Can you update to version 1.5.3

Let me try.

**Comment by RolandTwilt** [comment]
It no longer gives an error but my device doesn't function.

**Comment by RolandTwilt** [comment]
```yalm
2023-04-15 09:56:51.609 tiny TUYA: Pushing 'onCommandCallback' on to queue
2023-04-15 09:56:51.657 tiny TUYA: Processing 'onCommandCallback' message
2023-04-15 09:56:51.657 tiny TUYA: Acquiring GIL for 'onCommandCallback'
2023-04-15 09:56:51.657 tiny TUYA: Calling message handler 'onCommand' on 'module' type object.
2023-04-15 09:56:51.658 tiny TUYA: onCommand called for Device bf2c126721c2606655zryw Unit 1: Parameter 'Open', Level: 100', Color:
2023-04-15 09:56:51.658 tiny TUYA: Device ID: bf2c126721c2606655zryw
2023-04-15 09:56:51.658 tiny TUYA: nValue: 2
2023-04-15 09:56:51.658 tiny TUYA: sValue: None Type <class 'str'>
2023-04-15 09:56:51.658 tiny TUYA: LastLevel: 0
2023-04-15 09:56:51.608 Status: User: Roland (IP: 192.168.0.248) initiated a switch command (4602/Rolgordijnen/Open)
2023-04-15 09:56:52.641 tiny TUYA: Command send to tuya :bf2c126721c2606655zryw, {'commands': [{'code': 'mach_operate', 'value': 'FZ'}]}
2023-04-15 09:56:52.656 tiny TUYA: Update device value:bf2c126721c2606655zryw Unit: 1 sValue: Open nValue: 0 TimedOut=0
2023-04-15 09:56:52.656 tiny TUYA: Acquiring GIL for 'onCommandCallback'
2023-04-15 09:56:58.015 tiny TUYA: Pushing 'onHeartbeatCallback' on to queue
2023-04-15 09:56:58.066 tiny TUYA: Processing 'onHeartbeatCallback' message
2023-04-15 09:56:58.066 tiny TUYA: Acquiring GIL for 'onHeartbeatCallback'
2023-04-15 09:56:58.066 tiny TUYA: Calling message handler 'onHeartbeat' on 'module' type object.
2023-04-15 09:56:58.066 tiny TUYA: onHeartbeat called
2023-04-15 09:56:58.066 tiny TUYA: onHeartbeat called skipped
2023-04-15 09:56:58.066 tiny TUYA: Acquiring GIL for 'onHeartbeatCallback'
2023-04-15 09:57:08.037 tiny TUYA: Pushing 'onHeartbeatCallback' on to queue
2023-04-15 09:57:08.087 tiny TUYA: Processing 'onHeartbeatCallback' message
2023-04-15 09:57:08.087 tiny TUYA: Acquiring GIL for 'onHeartbeatCallback'
2023-04-15 09:57:08.087 tiny TUYA: Calling message handler 'onHeartbeat' on 'module' type object.
2023-04-15 09:57:08.088 tiny TUYA: onHeartbeat called
2023-04-15 09:57:08.088 tiny TUYA: onHeartbeat called skipped
2023-04-15 09:57:08.088 tiny TUYA: Acquiring GIL for 'onHeartbeatCallback'
```

**Comment by Xenomes** [comment]
Can you create a dump.json with debug_discovery.py and sent it to xenomes@outlook.com then can debug the issue.

**Comment by RolandTwilt** [comment]
Sure

**Comment by RolandTwilt** [comment]
> Can you create a dump.json with debug_discovery.py and sent it to [xenomes@outlook.com](mailto:xenomes@outlook.com) then can debug the issue.

Sent.

**Comment by Xenomes** [comment]
Fixed in the new version 1.5.4
